### ISSUE TITLE
MOSCAP simulation using fipy

### ISSUE THREAD

**Issue by Anweshan00005** [usnistgov/fipy#731]
I am trying to solve the following set of coupled pdes ( variables are p,n and psi) using fipy :-
`
qDn∇2n − qun∇.(n∇ψ) = q(n-10**11)/10**(-6), in Domain 1,
qDp∇2p + qup∇.(p∇ψ) = -q(p-10**21)/10**(-6), in Domain 1,
e∇2ψ = −(p − n- N) in Domain  1,
e∇2ψ = 0 in Domain 2, where Domain 1 is the region defined by y<=h and Domain 2 is the region defined by h<y<=h+tox
`
subjected to the boundary conditions :

```
p(x,0)= p0, n(x,0)= n0, psi(x,0)= 0, p(x,h)= p0*exp(-psi/Vth), n(x,h)= n0*exp(psi/Vth), psi(x,h+tox)= 5 
```
 I don't want to solve for the variables 'p' and 'n' in domain 2 but I want to solve for ''psi'' in both the domains. Since fipy does not allow explicit domain partition, I am using the internal constraint feature in fipy to impose the boundary conditions at the position given by `y=h`  , since this physical boundary  is actually inside the defined mesh i.e. it is not a mesh boundary.  I have already written the code for the simulation. The code has been attached below:
```python
from fipy import *
L= 5e-6
h= 10e-6
tox= 0.1*1e-6
q=1.6*1e-19
un=0.14
up=0.045
Vth=0.026
Dp= up*Vth
Dn=un*Vth
p0= 1e21
n0= 1e11
e0=8.854*1e-12

mesh1= Grid2D(dx= L/100,nx=100,dy=h/100,ny=100) # mesh for domain 1 for solving for p and n
mesh2= Grid2D(dx= L/100,nx=100,dy=tox/10,ny=10)
mesh3= mesh1+(mesh2+[[0],[h]]) # mesh for Domain 1 and Domain 2 for solving for psi

x = mesh3.cellCenters[0]
y = mesh3.cellCenters[1]


N= 1e21*(y<=h) # N changes from Domain 1 to Domain 2
e= 11.9*e0*(y<=h)+ 3.9*e0*(y>h) # e changes from Domain 1 to Domain 2


p=CellVariable(name='hole',mesh=mesh3,hasOld=True,value=p0)
n=CellVariable(name='electron',mesh=mesh3,hasOld=True,value=n0)
psi=CellVariable(name='potential',mesh=mesh3,hasOld=True,value=2.)

p.constrain(p0, where=mesh3.facesBottom)
n.constrain(n0, where=mesh3.facesBottom)
psi.constrain(0., where=mesh3.facesBottom)
psi.constrain(5., where=mesh3.facesTop)

mask1=((y==0)) # Boundary 1
mask2=((y>h - L/1000) * (y < h)) # Boundary2
mask3=(y==(h+tox)) # Boundary 3
largevalue= 1e45

eq1=(TransientTerm(var=n)+DiffusionTerm(coeff=q*Dn,var=n)-ConvectionTerm(coeff=q*un*psi.faceGrad,var=n)
     == ImplicitSourceTerm(coeff=q/1e-6,var=n)-q*n0/1e-6
     - ImplicitSourceTerm(coeff=mask2 * largevalue, var=n) + mask2 * largevalue * n0*numerix.exp(psi/Vth))
eq2=(TransientTerm(var=p)+DiffusionTerm(coeff=q*Dp,var=p)+ConvectionTerm(coeff=q*up*psi.faceGrad,var=p)
     ==ImplicitSourceTerm(coeff=-q/1e-6,var=p)+q*p0/1e-6
     - ImplicitSourceTerm(coeff=mask2 * largevalue, var=p) + mask2 * largevalue * p0*numerix.exp(-psi/Vth))
eq3=(TransientTerm(var=psi)+DiffusionTerm(coeff=e,var=psi)==-q*(p-n-N)*(y <= h)+0*(y>h))

eq= eq1 & eq2 & eq3

for i in range(100):
    p.updateOld()
    n.updateOld()
    psi.updateOld()
    res=1e10
    while res>1e-6:
        res=eq.sweep(dt=5) # a large time step dt=5 has been chosen because there is no transient term actually present in my pde.

viewer1=Matplotlib2DViewer(vars=p)
viewer2=Matplotlib2DViewer(vars=n)
viewer3=Matplotlib2DViewer(vars=psi)
viewer1.plot()
viewer2.plot()
viewer3.plot()
```

The above code runs without any error. But the results obtained are nowhere near the expected solution. Also, it is seen that the values of p,n and psi at the boundaries do not comply with the specified boundary conditions. Any help regarding this would be highly appreciated.
**Please note** that I am basically simulating a **MOSCAP(Semiconductor)** using the above code.

**Comment by guyer** [comment]
I don't have a solution for you, but I've posted a [notebook](https://gist.github.com/guyer/cefe482f409f3e31664f7e9e4e5acc78) which illustrates some of the steps I've found necessary over the years to solve these miserable equations.

The internal boundary condition you want is deeply non-linear and I've not yet found a way to make FiPy apply it. Smaller timesteps, finer meshing, more creative damping, or a moonlit sacrifice might help. Might not.

Since you're interested in steady-state solutions, you might have better luck with a pseudo-Fermi-level formulation.
